packages/visx-xychart/src/components/series/private/BaseBarStack.tsx

```diff
@@ -0,0 +1,237 @@
+  const { showTooltip, hideTooltip } = useContext(TooltipContext) ?? {};
+
+  const handleMouseMove = useCallback(
+    (params: HandlerParams | undefined) => {
+      const { svgPoint } = params || {};
+
+      // invoke showTooltip for each key so all data is available in context,
+      // and let Tooltip find the nearest point among them
+      dataKeys.forEach(key => {
+        const entry = dataRegistry.get(key);
+        const childData = barSeriesChildren.find(child => child.props.dataKey === key)?.props.data;
+        if (childData && svgPoint && width && height && showTooltip) {
+          const datum = findNearestStackDatum(
+            {
+              point: svgPoint,
+              data: entry.data,
+              xScale,
+              yScale,
+              xAccessor: entry.xAccessor,
+              yAccessor: entry.yAccessor,
+              width,
+              height,
+            },
+            childData,
+            horizontal,
+          );
+
+          if (datum) {
+            showTooltip({
+              key,
+              svgPoint,
+              ...datum,
+            });
+          }
+        }
+      });
+    },
+    [
+      barSeriesChildren,
+      dataRegistry,
+      dataKeys,
+      horizontal,
+      xScale,
+      yScale,
+      width,
+      height,
+      showTooltip,
+    ],
+  );
+  useEventEmitter('mousemove', handleMouseMove);
+  useEventEmitter('mouseout', hideTooltip);
+
+  // if scales and data are not available in the registry, bail
+  if (dataKeys.some(key => dataRegistry.get(key) == null) || !xScale || !yScale || !colorScale) {
+    return null;
+  }
+
+  const barThickness = getBandwidth(horizontal ? yScale : xScale);
+  const halfBarThickness = barThickness / 2;
+
+  let getWidth: (bar: StackBar) => number | undefined;
+  let getHeight: (bar: StackBar) => number | undefined;
+  let getX: (bar: StackBar) => number | undefined;
+  let getY: (bar: StackBar) => number | undefined;
+
+  if (horizontal) {
+    getWidth = bar => (xScale(getSecondItem(bar)) || 0) - (xScale(getFirstItem(bar)) || 0);
+    getHeight = () => barThickness;
+    getX = bar => xScale(getFirstItem(bar));
+    getY = bar =>
+      'bandwidth' in yScale
+        ? yScale(getStackValue(bar.data))
+        : Math.max((yScale(getStackValue(bar.data)) || 0) - halfBarThickness);
+  } else {
+    getWidth = () => barThickness;
+    getHeight = bar => (yScale(getFirstItem(bar)) || 0) - (yScale(getSecondItem(bar)) || 0);
+    getX = bar =>
+      'bandwidth' in xScale
+        ? xScale(getStackValue(bar.data))
+        : Math.max((xScale(getStackValue(bar.data)) || 0) - halfBarThickness);
+    getY = bar => yScale(getSecondItem(bar));
+  }
+
+  const bars = stackedData
+    .flatMap((barStack, stackIndex) => {
+      const entry = dataRegistry.get(barStack.key);
+      if (!entry) return null;
+
+      return barStack.map((bar, index) => {
+        const barX = getX(bar);
+        if (!isValidNumber(barX)) return null;
```

Comment:
should we log a warning here? or are there cases where devs would expect bar{X|Y} to not be a valid number?

Reply:
fair point. I have an item in the project tracker to do a sweep across series to make sure they handle `null`s / missing values properly so will think about it for that. my sense is that we don't need a warning since `null` values likely are common.